src/maxwell-web-view.js

```diff
@@ -0,0 +1,239 @@
+            maxwell.children[child.id] = child;
+
+            /* Collect children to allocate */
+            children.push(child.id);
+        }
+    }
+
+    /* Allocate GtkWidget, canvas size will change the first time
+     * child_draw() is called and we actually have something to show
+     */
+    window.webkit.messageHandlers.maxwell_children_allocate.postMessage(children);
+
+    /* Extra paranoid, update positions if anything changes in the DOM tree!
+     * ideally it would be nice to directly observe BoundingClientRect changes.
+     */
+    update_positions_throttled();
+};
+
+/* Main DOM observer */
+let observer = new MutationObserver(document_mutation_handler);
+observer.observe(document, {
+    childList: true,
+    subtree: true,
+    attributes: true
+});
+
+/* Get image data */
+function get_image (id, image_id, width, height) {
+    let xhr = new XMLHttpRequest();
+    let uri = 'maxwell:///' + id;
+
+    if (image_id)
+        uri += '?' + image_id;
+
+    xhr.open('GET', uri, false);
+    xhr.responseType = 'arraybuffer';
+
+    try {
+        xhr.send();
+    } catch (error) {
+        return null;
+    }
+
+    if (xhr.response) {
+        let data = new Uint8ClampedArray(xhr.response);
+        return new ImageData(data, width, height);
+    }
+
+    return null;
+}
+
+/* Semi Public API */
+
+/* Main entry point */
+window.maxwell = {};
+
+/* List of children */
+window.maxwell.children = {};
+
+/* child_resize()
+ */
+window.maxwell.child_resize = function (id, width, height) {
+    let child = maxwell.children[id];
+
+    if (!child || (child.width === width && child.height === height))
+        return;
+
+    /* Get image data first */
+    let image = get_image(id, null, width, height);
+
+    /* Resize canvas */
+    child.width = width;
+    child.height = height;
+
+    /* Update contents ASAP */
+    if (image)
+        child.getContext('2d').putImageData(image, 0, 0);
+}
+
+/* child_draw()
+ *
+ * Draw child canvas with maxwell:///canvasid image
+ *
+ * Unfortunatelly WebGL does not support texture_from_pixmap but we might be able
+ * to use GL to implement this function if we can get the context from WebKit
+ * itself
+ */
+window.maxwell.child_draw = function (id, image_id, x, y, width, height) {
+    let child = maxwell.children[id];
+
+    if (!child)
+        return;
+
+    /* Get image data */
+    let image = get_image(id, image_id, width, height);
+
+    /* Update contents */
+    if (image)
+        child.getContext('2d').putImageData(image, x, y);
+}
+
+/* child_set_visible()
+ *
+ * Show/hide widget element
+ */
+window.maxwell.child_set_visible = function (id, visible) {
+    let child = maxwell.children[id];
+
+    if (child)
+        child.style.display = (visible) ? child.maxwell_display_value : 'none';
+}
+
+/* child_init()
+ *
+ */
+window.maxwell.child_init = function (id, width, height, visible) {
+    let child = maxwell.children[id];
+
+    if (!child)
+        return;
```

Comment:
All the functions that this calls already check that `maxwell.children[id]` is not undefined. Since `child` here is not used directly, perhaps we can remove this code.